### PR TITLE
handle concurrent requests 

### LINKED ISSUE #1939: Relaxed concurrency in poolmanager strategy
Due to the current implementation of cache(pool cache) for the pool manager, sometimes number of specialized pods (function pods) are more than the `concurrency` parameter, this happens when there is burst of requests and number of incoming requests is greater than `concurrency` parameter.

Solution
This can be improved by applying locking mechanism but would degrade performance of the whole specialization.

### PR DESCRIPTION
**Pull request by pranoyk**
## Description
Keep a track of number of pods in specialization and make check the capacity is more then requested else fail the function

### Terminology
- Capacity - (specialized pod + pods in specialization state) * rpp - (total active requests + functions awaiting svc)
- svcWaiting - number of functions waiting for a svc
  
### Approach
- During getSvcValue if svc does not exist increase svcWaiting counter and check if capacity is greater then 1 (In the case of first request capacity would be less then 1)
- If capacity is greater then 1 then push req into the queue and await specialization to finish
- once specialization is done check if req are awaiting in queue for this function svc. If yes then allocate the svc to functions upto its requests per pod limit
- keep repeating the process until the limit is reached.

Fixes #1939 

## Test scenarios
<div dir="ltr" style="margin-left:0pt;" align="left" id="docs-internal-guid-7bb3e63e-7fff-293f-ecf7-50ea1e03b279">

RPP | Concurrency | Requests | Expected | Actual
-- | -- | -- | -- | --
60 | 5 | 300 | Specialized Pods - 5 Requests Success - 300 Requests Fail - 0 | Specialized Pods - 5 Requests Success - 300 Requests Fail - 0
60 | 5 | 300 concurrent requests for 60 sec | Specialized Pods - 5 Requests Success - ~18000 Requests Fail - 0 | Specialized Pods - 5 Requests Success - 14464 Requests Fail - 0 No of requests made can vary due to the test tool used (ab in this case)
1 | 10 | 10 | Specialized Pods - 10 Requests Success - 10 Requests Fail - 0 | Specialized Pods - 10 Requests Success - 10 Requests Fail - 0
5 | 5 | 25 | Specialized Pods - 5 Requests Success - 25 Requests Fail - 0 | Specialized Pods - 5 Requests Success - 25 Requests Fail - 0
5 | 1 | 5 | Specialized Pods - 1 Requests Success - 5 Requests Fail - 0 | Specialized Pods - 1 Requests Success - 5 Requests Fail - 0
5 | 1 | 6 | Specialized Pods - 1 Requests Success - 5 Requests Fail - 1 | Specialized Pods - 1 Requests Success - 5 Requests Fail - ~1 (requests gets timed out) In case the time taken to run the function is not very long then all 6 requests may be success
1 | 5 | 6 | Specialized Pods - 5 Requests Success - 5 Requests Fail - 1 | Specialized Pods - 5 Requests Success - 5 Requests Fail - 1 (requests gets timed out) In case the time taken to run the function is not very long then all 6 requests may be success
5 | 5 | 10 | Specialized Pods -  =<5 Requests Success - 10 Requests Fail - 0 | Specialized Pods - =<5 Requests Success - 10 Requests Fail - 0
10 | 1 | 20 | Specialized Pods - 1 Requests Success - 10 Requests Fail - 10 | Specialized Pods - 1 Requests Success - 10 Requests Fail - 10

</div>

### CODE REVIEW

File: pkg/executor/executortype/newdeploy/newdeploymgr.go [hunk 1]
```diff
@@ -200,11 +200,14 @@ func (deploy *NewDeploy) UnTapService(ctx context.Context, key string, svcHost s
...
 // GetFuncSvcFromPoolCache has not been implemented for NewDeployment
-func (deploy *NewDeploy) GetFuncSvcFromPoolCache(ctx context.Context, fn *fv1.Function, requestsPerPod int) (*fscache.FuncSvc, int, error) {
+func (deploy *NewDeploy) GetFuncSvcFromPoolCache(ctx context.Context, fn *fv1.Function, requestsPerPod int, concurrency int) (*fscache.FuncSvc, error) {
 	// Not Implemented for NewDeployment. Will be used when support of concurrent specialization of same function is added.
-	return nil, 0, nil
+	return nil, nil
 }
 
+// SpecializationStart has not been implemented for NewDeployment
+func (deploy *NewDeploy) SpecializationStart(context.Context, *fv1.Function) {}
```
Comment (sanketsudake) [inline pkg/executor/executortype/newdeploy/newdeploymgr.go:209]: Please remove this

**Review by sanketsudake — CHANGES_REQUESTED** [review]

File: pkg/apis/core/v1/types.go [hunk 2]
```diff
@@ -863,3 +863,17 @@ type (
 func (a Archive) IsEmpty() bool {
 	return len(a.Literal) == 0 && len(a.URL) == 0
 }
+
+func (fn Function) GetConcurrency() int {
+	if fn.Spec.Concurrency == 0 {
+		return 500
```
Comment (sanketsudake) [inline pkg/apis/core/v1/types.go:869]: Please define constants above and use them here.

DefaultConcurrency and DefaultRequestPerPod

File: pkg/executor/executortype/container/containermgr.go [hunk 3]
```diff
@@ -194,9 +194,9 @@ func (caaf *Container) DeleteFuncSvcFromCache(ctx context.Context, fsvc *fscache
 }
 
 // GetFuncSvcFromPoolCache has not been implemented for Container Functions
-func (caaf *Container) GetFuncSvcFromPoolCache(ctx context.Context, fn *fv1.Function, requestsPerPod int) (*fscache.FuncSvc, int, error) {
+func (caaf *Container) GetFuncSvcFromPoolCache(ctx context.Context, fn *fv1.Function) (*fscache.FuncSvc, error) {
```
[the remark this answers is not in the document]
Comment (sanketsudake) [reply pkg/executor/executortype/container/containermgr.go:197]: Remove method from interface as well.

File: pkg/executor/executortype/newdeploy/newdeploymgr.go [hunk 4]
```diff
@@ -200,9 +200,9 @@ func (deploy *NewDeploy) UnTapService(ctx context.Context, key string, svcHost s
 }
 
 // GetFuncSvcFromPoolCache has not been implemented for NewDeployment
-func (deploy *NewDeploy) GetFuncSvcFromPoolCache(ctx context.Context, fn *fv1.Function, requestsPerPod int) (*fscache.FuncSvc, int, error) {
+func (deploy *NewDeploy) GetFuncSvcFromPoolCache(ctx context.Context, fn *fv1.Function) (*fscache.FuncSvc, error) {
```
Comment (sanketsudake) [inline pkg/executor/executortype/newdeploy/newdeploymgr.go:203]: Please remove method.

File: pkg/executor/fscache/poolcache.go [hunk 5]
```diff
@@ -88,52 +98,117 @@ func NewPoolCache(logger *zap.Logger) *PoolCache {
...
+				continue
+			}
+			// concurrency should not be set to zero and
+			//sum of sprcialization in progress and specialized pods should be less then req.concurrency
+			if req.concurrency > 0 && len(funcSvcGroup.svcs)+funcSvcGroup.specializationInProgress > req.concurrency {
+				resp.error = ferror.MakeError(ferror.ErrorTooManyRequests, fmt.Sprintf("function '%s' concurrency '%d' limit reached", req.function, req.concurrency))
+			} else {
+				funcSvcGroup.svcWaiting++
+				resp.error = ferror.MakeError(ferror.ErrorNotFound, fmt.Sprintf("function '%s' all functions are busy", req.function))
+				if capacity > 0 {
```
Comment (sanketsudake) [inline pkg/executor/fscache/poolcache.go:151]: We are not checking capacity > 1 ?

File: pkg/executor/fscache/poolcache.go [hunk 6]
```diff
@@ -88,52 +98,117 @@ func NewPoolCache(logger *zap.Logger) *PoolCache {
...
+				svcCapacity := req.requestsPerPod - c.cache[req.function].svcs[req.address].activeRequests
+				queueLen := c.cache[req.function].queue.Len()
+				if svcCapacity > queueLen {
+					svcCapacity = queueLen
+				}
+				for i := 0; i < svcCapacity; i++ {
+					popped := c.cache[req.function].queue.Pop()
+					if popped != nil {
+						if popped.ctx.Err() == nil {
+							popped.svcChannel <- req.value
```
Comment (sanketsudake) [inline pkg/executor/fscache/poolcache.go:181]: Close channel once value is sent ?

File: pkg/executor/fscache/poolcache.go [hunk 7]
```diff
@@ -88,52 +98,117 @@ func NewPoolCache(logger *zap.Logger) *PoolCache {
...
+				c.cache[req.function].svcWaiting--
+				svcCapacity := req.requestsPerPod - c.cache[req.function].svcs[req.address].activeRequests
+				queueLen := c.cache[req.function].queue.Len()
+				if svcCapacity > queueLen {
+					svcCapacity = queueLen
+				}
+				for i := 0; i < svcCapacity; i++ {
+					popped := c.cache[req.function].queue.Pop()
+					if popped != nil {
+						if popped.ctx.Err() == nil {
```
Comment (sanketsudake) [inline pkg/executor/fscache/poolcache.go:180]: If requests has timed out we can drop request. And consider additional in place ?
eg. top of queue has 4 timed out request then we have capacity of 4 then we should drop first 4 and consider next 4 ? 

Also its quite possible that we can run into race condition. Just need to ensure we are not working on stale capacity value.

File: pkg/executor/fscache/poolcache.go [hunk 8]
```diff
@@ -88,48 +97,99 @@ func NewPoolCache(logger *zap.Logger) *PoolCache {
...
+			if c.cache[req.function].svcWaiting > 0 {
+				c.cache[req.function].svcWaiting--
+				svcCapacity := req.requestsPerPod - c.cache[req.function].svcs[req.address].activeRequests
+				queueLen := c.cache[req.function].queue.Len()
+				if svcCapacity > queueLen {
+					svcCapacity = queueLen
+				}
+				for i := 0; i <= svcCapacity; {
+					popped := c.cache[req.function].queue.Pop()
+					if popped != nil {
```
Comment (sanketsudake) [inline pkg/executor/fscache/poolcache.go:180]: Invert condition
```
if  popped == nil {
      break
}
... rest of code
```
for better readability.

File: pkg/executor/fscache/poolcache.go [hunk 9]
```diff
@@ -36,6 +36,8 @@ const (
 	markAvailable
 	deleteValue
 	setCPUUtilization
+	specializationStart
```
Comment (sanketsudake) [inline pkg/executor/fscache/poolcache.go:39]: We can remove these and related code if not needed anymore.

**Comment by LyKos4** [comment]
When I set concurrency 1 or rpp 1 and try to run a  function more than 1 pods are created for that run, why does this happen?
Same for yolo true

Doesn't yolo true of rpp 1 mean that a function run should be done by a new single pod, meaning 1 pod per function run?

**Pull request merged**